File: inventory-framework-platform-bukkit/src/main/java/me/devnatan/inventoryframework/bukkit/pipeline/GlobalClickInterceptor.java
package me.devnatan.inventoryframework.bukkit.pipeline;

import me.devnatan.inventoryframework.api.VirtualView;
import me.devnatan.inventoryframework.api.pipeline.PipelineContext;
import me.devnatan.inventoryframework.api.pipeline.PipelineInterceptor;
import me.devnatan.inventoryframework.bukkit.context.SlotClickContext;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.jetbrains.annotations.NotNull;

import static me.devnatan.inventoryframework.api.ViewConfig.CANCEL_ON_CLICK;

/**
 * Intercepted when a player clicks on the view container.
 * If the click is canceled, this interceptor ends the pipeline immediately.
 */
public final class GlobalClickInterceptor implements PipelineInterceptor<VirtualView> {

	@Override
	public void intercept(@NotNull PipelineContext<VirtualView> pipeline, @NotNull VirtualView subject) {
		if (!(subject instanceof SlotClickContext)) return;

		final SlotClickContext context = (SlotClickContext) subject;
		final InventoryClickEvent event = context.getClickOrigin();

		// inherit cancellation so we can un-cancel it
		context.setCancelled(event.isCancelled() || context.getConfig().isOptionSet(CANCEL_ON_CLICK, true));
		context.getRoot().onClick(context);
	}
}

File: inventory-framework-platform-minestom/src/main/kotlin/me/devnatan/inventoryframework/minestom/pipeline/GlobalClickInterceptor.kt
package me.devnatan.inventoryframework.minestom.pipeline

import me.devnatan.inventoryframework.api.ViewConfig
import me.devnatan.inventoryframework.api.VirtualView
import me.devnatan.inventoryframework.api.pipeline.PipelineContext
import me.devnatan.inventoryframework.api.pipeline.PipelineInterceptor
import me.devnatan.inventoryframework.minestom.context.SlotClickContext
import net.minestom.server.event.inventory.InventoryPreClickEvent

/**
 * Intercepted when a player clicks on the view container. If the click is canceled, this
 * interceptor ends the pipeline immediately.
 */
class GlobalClickInterceptor : PipelineInterceptor<VirtualView> {
    override fun intercept(
        pipeline: PipelineContext<VirtualView>,
        subject: VirtualView,
    ) {
        if (subject !is SlotClickContext) return

        val event: InventoryPreClickEvent = subject.clickOrigin

        // inherit cancellation so we can un-cancel it
        subject.isCancelled =
            event.isCancelled || subject.config.isOptionSet(ViewConfig.CANCEL_ON_CLICK, true)
        subject.root.onClick(subject)
    }
}
